jib-core/src/main/java/com/google/cloud/tools/jib/filesystem/XdgDirectories.java

```diff
@@ -0,0 +1,161 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.filesystem;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Properties;
+import java.util.logging.Logger;
+
+/**
+ * Obtains OS-specific directories based on the XDG Base Directory Specification.
+ *
+ * <p>Specifically, from the specification:
+ *
+ * <ul>
+ *   <li>These directories are defined by the environment variables {@code $XDG_CACHE_HOME} and
+ *       {@code $XDG_CONFIG_HOME}.
+ *   <li>If {@code $XDG_CACHE_HOME} / {@code $XDG_CONFIG_HOME} is either not set or empty, a
+ *       platform-specific equivalent of {@code $HOME/.cache} / {@code $HOME/.config} should be
+ *       used.
+ * </ul>
+ *
+ * @see <a
+ *     href="https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html">https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html</a>
+ */
+public class XdgDirectories {
+
+  private static final Logger LOGGER = Logger.getLogger(XdgDirectories.class.getName());
+  private static final Path JIB_SUBDIRECTORY_LINUX =
+      Paths.get("google-cloud-tools-java").resolve("jib");
+  private static final Path JIB_SUBDIRECTORY_OTHER = Paths.get("Google").resolve("Jib");
+
+  public static Path getCacheHome() {
+    return getCacheHome(System.getProperties(), System.getenv());
+  }
+
+  public static Path getConfigHome() {
+    return getConfigHome(System.getProperties(), System.getenv());
+  }
+
+  /**
+   * Returns {@code $XDG_CACHE_HOME}, if available, or resolves the OS-specific user cache home
+   * based.
+   *
+   * <p>For Linux, this is {@code $HOME/.cache/google-cloud-tools-java/jib/}.
+   *
+   * <p>For Windows, this is {@code %LOCALAPPDATA%\Google\Jib\Cache\}.
+   *
+   * <p>For macOS, this is {@code $HOME/Library/Caches/Google/Jib/}.
+   */
+  @VisibleForTesting
+  static Path getCacheHome(Properties properties, Map<String, String> environment) {
+    return getOsSpecificDirectory(
+        properties, environment, "XDG_CACHE_HOME", ".cache", "Cache", "Caches");
+  }
+
+  /**
+   * Returns config directory based on {@code $XDG_CONFIG_HOME}, if available, or resolves the
+   * OS-specific user config directory.
+   *
+   * <p>For Linux, this is {@code $HOME/.config/google-cloud-tools-java/jib/}.
+   *
+   * <p>For Windows, this is {@code %LOCALAPPDATA%\Google\Jib\Config\}.
+   *
+   * <p>For macOS, this is {@code $HOME/Library/Preferences/Google/Jib/}.
+   */
+  @VisibleForTesting
+  static Path getConfigHome(Properties properties, Map<String, String> environment) {
+    return getOsSpecificDirectory(
+        properties, environment, "XDG_CONFIG_HOME", ".config", "Config", "Preferences");
```

Comment:
This might end up being a pain, since the function operates on portions of the os-specific subdirectories, not the whole thing (e.g. mac needs to check if `~/Library/Preferences` exists ignoring the `JIB_SUBDIRECTORY_OTHER` portion, if `$XDG_X_HOME` isn't configured then each OS defaults to `$HOME/.config/[OS-specific subdirectory], etc.)